File: python3/BoundaryofBinaryTree2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def boundaryOfBinaryTree(self, root: TreeNode) -> List[int]:
        if root == None:
            return []
        
        output = [root.val]
        if root.left == None and root.right == None:
            return output 
        
        left = []
        right = []
        leaves = []
        
        ROOT = 0
        LEFT = 1
        RIGHT = 2
        INNER = 3
        
        def dfs(curNode, flag):
            if (curNode == None):
                return
            
            if curNode.left == None and curNode.right == None:
                leaves.append(curNode.val)
                return
            
            if flag == LEFT:
                left.append(curNode.val)
            elif flag == RIGHT:
                right.append(curNode.val)

            newFlag = -1
            if flag == ROOT or flag == LEFT:
                newFlag = LEFT
            elif flag == RIGHT:
                if curNode.right == None:
                    newFlag = RIGHT
                else:
                    newFlag = INNER
            dfs(curNode.left, newFlag)
            
            if flag == ROOT or flag == RIGHT:
                newFlag = RIGHT
            elif flag == LEFT:
                if curNode.left == None:
                    newFlag = LEFT
                else:
                    newFlag = INNER
            dfs(curNode.right, newFlag)
            
            return
        
        dfs(root, ROOT)
        
        for item in left:
            output.append(item)
        
        for item in leaves:
            output.append(item)
        
        for i in range(len(right)-1, -1, -1):
            output.append(right[i])
                   
        return output

File: c++/BoundaryofBinaryTree2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#define ROOT 0
#define LEFT 1
#define RIGHT 2
#define INNER 3

class Solution {

private:
        vector<int> left;
        vector<int> leaves;
        vector<int> right;
    
public:
    void dfs(TreeNode* curNode, int flag) {
        if (curNode == NULL)
            return;

        if (curNode->left == NULL && curNode->right == NULL) {
            leaves.push_back(curNode->val);
            return;
        }

        if (flag == LEFT)
            left.push_back(curNode->val);
        else if (flag == RIGHT)
            right.push_back(curNode->val);

        int newFlag = -1;
        if (flag == ROOT || flag == LEFT)
            newFlag = LEFT;
        else if (flag == RIGHT) {
            if (curNode->right == NULL) {
                newFlag = RIGHT;
            } else {
                newFlag = INNER;
            }
        }
        dfs(curNode->left, newFlag);

        if (flag == ROOT || flag == RIGHT)
            newFlag = RIGHT;
        else if (flag == LEFT) {
            if (curNode->left == NULL)
                newFlag = LEFT;
            else
                newFlag = INNER;
        }
        dfs(curNode->right, newFlag);
        return;
    }
    
    vector<int> boundaryOfBinaryTree(TreeNode* root) {
        vector<int> output;
        if (root == NULL)
            return output;
        
        output.push_back(root->val);
        if (root->left == NULL && root->right == NULL)
            return output;
                
        dfs(root, ROOT);
        
        for (auto item : left)
            output.push_back(item);
        
        for (auto item : leaves)
            output.push_back(item);
        
        for (int i = right.size() - 1; i >= 0; --i)
            output.push_back(right[i]);
                   
        return output;
    }
};
